key_cacher.go

```diff
@@ -1,101 +1,108 @@
 package auth0
 
 import (
+	"errors"
 	"time"
 
 	jose "gopkg.in/square/go-jose.v2"
 )
 
+var (
+	ErrNoKeyFound  = errors.New("no Keys has been found")
+	ErrKeyExpired  = errors.New("key exists but is expired")
+	MaxAgeNoCheck  = time.Duration(-1)
+	MaxSizeNoCheck = -1
+)
+
 type KeyCacher interface {
-	Get(keyID string) (jose.JSONWebKey, bool)
-	Add(keyID string, webKeys []jose.JSONWebKey) (jose.JSONWebKey, bool)
+	Get(keyID string) (*jose.JSONWebKey, error)
+	Add(keyID string, webKeys []jose.JSONWebKey) (*jose.JSONWebKey, error)
 }
 
 type memoryKeyCacher struct {
 	entries map[string]keyCacherEntry
-	maxAge  int
-	size    int
+	maxAge  time.Duration
+	maxSize int
 }
 
 type keyCacherEntry struct {
 	addedAt time.Time
 	jose.JSONWebKey
 }
 
-func NewMemoryKeyCacher(maxAge int, size int) KeyCacher {
+func NewMemoryKeyCacher(maxAge time.Duration, maxSize int) KeyCacher {
 	return &memoryKeyCacher{
 		entries: map[string]keyCacherEntry{},
```

Comment:
VERY IMPORTANT, should we move the mutex protecting the map in the cacher or not?